File: compiler-d8/src/main/java/com/xiaoyv/java/compiler/tools/java/JavaClassHelper.java
package com.xiaoyv.java.compiler.tools.java;

import androidx.annotation.NonNull;

import com.xiaoyv.java.compiler.utils.FileIOUtils;
import com.xiaoyv.java.compiler.utils.FileUtils;

import java.io.File;
import java.util.List;
import java.util.regex.Pattern;

/**
 * @author 王怀玉
 * @date 18/07/2017
 */

public class JavaClassHelper {
    private static final Pattern MAIN_FUNCTION;

    private static final String MAIN_FUN_REGEX =
            "(public\\s+static\\s+void\\s+main\\s?)" +
                    "(\\(\\s?String\\s?\\[\\s?]\\s?\\w+\\s?\\)|" +
                    "(\\(\\s?String\\s+\\w+\\[\\s?]\\s?\\)))";

    static {
        MAIN_FUNCTION = Pattern.compile(MAIN_FUN_REGEX, Pattern.DOTALL);
    }

    /**
     * 检测 代码是否含有 public static void main(String[] args) {}方法
     */
    private static boolean matchMainFunction(String content) {
        return MAIN_FUNCTION.matcher(content).find();
    }

    /**
     * 检测 java 文件是否有主方法
     */
    public static boolean hasMainFunction(File javaFile) {
        String s = FileIOUtils.readFile2String(javaFile);
        return matchMainFunction(s);
    }

    /**
     * 检测 java 文件是否有主方法
     */
    public static boolean hasMainFunction(String javaFilePath) {
        String s = FileIOUtils.readFile2String(javaFilePath);
        return matchMainFunction(s);
    }


    /**
     * 获取当前项目 lib 目录的 jar 文件的 CLASSPATH
     *
     * @param libFolder 当前项目添加的jar依赖文件夹
     * @return 所有依赖转成的 -classpath 格式字符串
     */
    public static String getLibClassPath(@NonNull File libFolder) {
        boolean existsDir = FileUtils.createOrExistsDir(libFolder);

        if (!existsDir) {
            return "";
        }

        List<File> javaLibraries = FileUtils.listFilesInDirWithFilter(libFolder, pathname ->
                FileUtils.isFile(pathname) && FileUtils.getFileName(pathname).endsWith(".jar"));

        if (javaLibraries == null) {
            return "";
        }

        StringBuilder classpath = new StringBuilder(".");
        for (File javaLibrary : javaLibraries) {
            if (classpath.length() != 0) {
                classpath.append(File.pathSeparator);
            }
            classpath.append(javaLibrary.getAbsolutePath());
        }
        return classpath.toString();
    }

}

File: compiler-d8/src/main/java/com/xiaoyv/java/compiler/tools/java/JavaClassHelper.kt
package com.xiaoyv.java.compiler.tools.java

import android.util.Log
import com.xiaoyv.java.compiler.utils.FileUtils
import java.io.File
import java.util.*
import java.util.jar.JarEntry
import java.util.jar.JarFile

/**
 * @author 王怀玉
 * @date 18/07/2017
 */
object JavaClassHelper {

    /**
     * 获取当前项目 lib 目录的 jar 文件的 CLASSPATH
     *
     * @param libFolder 当前项目添加的jar依赖文件夹
     * @return 所有依赖转成的 -classpath 格式字符串
     */
    @JvmStatic
    fun getLibClassPath(libFolder: File): String {
        val existsDir = FileUtils.createOrExistsDir(libFolder)
        if (!existsDir) {
            return ""
        }

        val javaLibraries = FileUtils.listFilesInDirWithFilter(libFolder) { file ->
            FileUtils.isFile(file) && FileUtils.getFileName(file).endsWith(".jar")
        } ?: return ""

        val classpath = StringBuilder(".")
        for (javaLibrary in javaLibraries) {
            if (classpath.isNotEmpty()) {
                classpath.append(File.pathSeparator)
            }
            classpath.append(javaLibrary.absolutePath)
        }
        return classpath.toString()
    }

    @JvmStatic
    fun getClasses(jarPath: String) = getClasses(File(jarPath))

    @JvmStatic
    fun getClasses(file: File): List<Class<*>> {
        if (file.exists().not()) {
            return arrayListOf()
        }
        val classes: MutableList<Class<*>> = arrayListOf()
        runCatching {
            val jarFile = JarFile(file)
            val entries: Enumeration<JarEntry> = jarFile.entries()
            while (entries.hasMoreElements()) {
                val jarEntry: JarEntry = entries.nextElement()
                if (jarEntry.isDirectory ||
                    jarEntry.name.endsWith(".class").not() ||
                    jarEntry.name.contains("$")
                ) {
                    continue
                }
                var className = jarEntry.name.substring(0, jarEntry.name.length - 6)
                className = className.replace('/', '.')
                runCatching {
                    classes.add(ClassLoader.getSystemClassLoader().loadClass(className))
                }.onFailure {
                    Log.e("JavaClassHelper", "编译器未适配该类： $className")
                }
            }
        }
        return classes
    }
}